sphinx_gallery/gen_rst.py

```diff
@@ -604,58 +600,113 @@ def generate_file_rst(fname, target_dir, src_dir, gallery_conf):
+    executed_blocks : list
 
-            time_elapsed += rtime
+    gallery_conf : dict
+        Sphinx-Gallery configuration dictionary
+
+    Returns
+    -------
+    out : str
+        RST notebook
+    """
+
+    # A simple example has two blocks: one for the
+    # example introduction/explanation and one for the code
+    is_example_notebook_like = len(script_blocks) > 2
+    example_rst = u""  # there can be unicode content
+    for (blabel, bcontent, lineno), code_output in zip(script_blocks, output_blocks):
+        if blabel == 'code':
 
             if not file_conf.get('line_numbers',
                                  gallery_conf.get('line_numbers', False)):
                 lineno = None
 
+            code_rst = codestr2rst(bcontent, lineno=lineno) + '\n'
             if is_example_notebook_like:
-                example_rst += codestr2rst(bcontent, lineno=lineno) + '\n'
+                example_rst += code_rst
                 example_rst += code_output
             else:
                 example_rst += code_output
                 if 'sphx-glr-script-out' in code_output:
                     # Add some vertical space after output
                     example_rst += "\n\n|\n\n"
-                example_rst += codestr2rst(bcontent, lineno=lineno) + '\n'
-
+                example_rst += code_rst
         else:
-            example_rst += bcontent + '\n\n'
+            example_rst += bcontent + '\n'
+    return example_rst
 
-    sys.argv = argv_orig
-    clean_modules()
 
-    # Writes md5 checksum if example has build correctly
-    # not failed and was initially meant to run(no-plot shall not cache md5sum)
-    if block_vars['execute_script']:
-        with open(example_file + '.md5', 'w') as file_checksum:
-            file_checksum.write(get_md5sum(example_file))
+def save_rst_notebook(example_rst, write_file, time_elapsed, gallery_conf):
+    """Saves the rst notebook to write_file including necessary header & footer
 
-    save_thumbnail(image_path_template, src_file, file_conf, gallery_conf)
+    Parameters
+    ----------
+    example_rst : str
+        rST containing the executed file content
 
-    time_m, time_s = divmod(time_elapsed, 60)
-    example_nb = jupyter_notebook(script_blocks)
-    save_notebook(example_nb, example_file.replace('.py', '.ipynb'))
-    with codecs.open(os.path.join(target_dir, base_image_name + '.rst'),
-                     mode='w', encoding='utf-8') as f:
-        if time_elapsed >= gallery_conf["min_reported_time"]:
-            example_rst += ("**Total running time of the script:**"
-                            " ({0: .0f} minutes {1: .3f} seconds)\n\n".format(
-                                time_m, time_s))
-        example_rst += CODE_DOWNLOAD.format(fname,
-                                            fname.replace('.py', '.ipynb'))
-        example_rst += SPHX_GLR_SIG
-        f.write(example_rst)
+    write_fname : str
+        Filename with full path where to save the rST file
```

Comment:
I think this is either rst (as in the extension) or reST (short name for reSrucuturedText). I'd rather use the extension, less change to be confused with REST. This may happen in other places in this PR, please check.